### PR TITLE
Dependent Dropdown/Checkbox Bug Fix

### PR DESCRIPTION
**Pull request by Jtang-1**
## Product Description

Example of bug: 
The bug being that the item at index 1 is chosen for "Gym". After "City" value changes and "Gym" values are repopulated, the new item at index 1 is selected even though it's not the same value as before.

https://user-images.githubusercontent.com/88759246/234439512-7b25213f-513f-438b-83a9-17533742af7e.mov

Corrected behavior:

https://user-images.githubusercontent.com/88759246/234442572-4d33bb93-fe05-4dcf-964b-b2494757810f.mov

## Technical Summary

[Spec](https://docs.google.com/document/d/1Xtly4YfrdaB1w1Ozwlwy9LQ4JIrVRmhvg-44-Rl-z6A/edit#)
Jira Ticket: https://dimagi-dev.atlassian.net/browse/USH-2292

This change has a `select_values_by_key` flag that will be passed from formplayer after formplayer change is deployed. https://github.com/dimagi/formplayer/pull/1362. Until then, the current behavior (selection via index) will continue.

## Feature Flag

Case Search

### Safety story

Local testing with and without new formplayer changes. 

Change does not affect data or app build and is limited to case search with dropdown/checkbox fields. Defect consequences would possibly be inability to run case search or returning incorrect results. The latter is not likely as Formplayer catches if the value (index or key) is not expected. 
### Automated test coverage

Added tests that key value pairs are set up correctly.

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-4993

### Rollback instructions

Formplayer changes https://github.com/dimagi/formplayer/pull/1362 will need to be rolled back first.

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 1]
```diff
@@ -226,6 +229,26 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
             'click @ui.searchForBlank': 'toggleBlankSearch',
         },
 
+        _setItemSet: function(itemsetChoicesLabels, itemsetChoiceskey) {
```
Comment (orangejenny) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:232]:
```suggestion
        _setItemset: function(itemsetChoicesLabels, itemsetChoicesKey) {
```

Comment (Jtang-1) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js]: [6aab589](https://github.com/dimagi/commcare-hq/pull/32883/commits/6aab58990341bd2b5b71336c4a4b66afd85df4be)

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 2]
```diff
@@ -555,7 +591,7 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
             var self = this;
 
             var urlObject = formplayerUtils.currentUrlToObject();
-            urlObject.setQueryData(self.getAnswers(), false);
+            urlObject.setQueryData(self.getAnswers(), false, false, selectValuesByKeys=self.selectValuesByKeys);
```
Comment (orangejenny) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:594]: Feel free to call out of scope for this PR, but named parameters for all the options to `setQueryData` would be great.

**Review by orangejenny — APPROVED** [review]

**Review by nospame — COMMENTED** [review]
A couple of questions/comments from pair code review with @gherceg

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js (hunk 2 above)
Comment (Jtang-1) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js]: [ac6cda5](https://github.com/dimagi/commcare-hq/pull/32883/commits/ac6cda562f9e7df2d8444ce79c2e090c2157e84a)

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 3]
```diff
@@ -226,6 +229,26 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
             'click @ui.searchForBlank': 'toggleBlankSearch',
         },
 
+        _setItemset: function(itemsetChoices, itemsetChoicesKey) {
+            itemsetChoices = itemsetChoices || [];
+            let itemsetChoicesDict = {};
+
+            if (this.parentView.selectValuesByKeys){
+                itemsetChoicesKey = itemsetChoicesKey || [];
+                itemsetChoicesKey.forEach((key,i) => itemsetChoicesDict[key]=itemsetChoices[i])
```
[the remark this answers is not in the document]
Comment (Jtang-1) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:238]: No, formplayer will provide a corresponding key for each choice so lengths will be the same.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 4]
```diff
@@ -413,6 +436,17 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
 
         initialize: function (options) {
             this.parentModel = options.collection.models;
+
+            // whether the select prompt selection is passed as itemset keys
+            // only here to maintain backward compatibility and can be removed
+            // once web apps fully transition using keys to convey select prompt selection.
+            this.selectValuesByKeys = false;
+            for (key in this.parentModel) {
+                if ("itemsetChoicesKey" in this.parentModel[key].attributes){
```
[the remark this answers is not in the document]
Comment (Jtang-1) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js]: Good catch, it's actually in index not a key. I also read for-in isn't intended to be used for looping through arrays so changed it to the old-fashioned way
[ac6cda5](https://github.com/dimagi/commcare-hq/pull/32883/commits/ac6cda562f9e7df2d8444ce79c2e090c2157e84a)

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 5]
```diff
@@ -226,6 +229,26 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
...
+
+            if (this.parentView.selectValuesByKeys){
+                itemsetChoicesKey = itemsetChoicesKey || [];
+                itemsetChoicesKey.forEach((key,i) => itemsetChoicesDict[key]=itemsetChoices[i])
+                this.model.set({
+                    itemsetChoicesKey: itemsetChoicesKey,
+                })
+            }
+            else{
+                itemsetChoices.forEach((key,i) => itemsetChoicesDict[i]=itemsetChoices[i])
```
[the remark this answers is not in the document]
Comment (Jtang-1) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js]: Yes it is, 
[e76f5c5](https://github.com/dimagi/commcare-hq/pull/32883/commits/e76f5c524370f63966d0ff874f52a4032262eabf)

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js]: Is `parentModel` an array? I would have imagined it was an object. I think you could probably also use a `for...of` loop to access the items directly if you wanted.

[the remark this answers is not in the document]
Comment (Jtang-1) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js]: Yes you're right, I ended up using for...of. I only just learned that was an option

 [7f003ac](https://github.com/dimagi/commcare-hq/pull/32883/commits/7f003ac9c008caddf2c56ed3ff6c535f681663a4)

**Pull request merged**